lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+      FRG->connectNodes(BBI->LastNode, SBI->FirstNode, C, D, isBackEdge);
+    }
+  }
+  assert(StructInfo->getExecutionCount(&F->getEntryBlock()));
+  auto *BBI = FRG->getBasicBlockHelperInfo(&F->getEntryBlock());
+  FRG->setRootNode(BBI->FirstNode);
+  DEBUG(dbgs() << "---------- FRG for function " << F->getName()
+               << "----------------\n");
+  DEBUG(FRG->debugPrint(dbgs()));
+  DEBUG(dbgs() << "------------------------------------------------------------"
+                  "------------\n");
+  return FRG;
+}
+
+// Core function to update a cell in CPT.
+// In this function, a CPT can be either CloseProximityTable (real CP relations)
+// or GoldCPT (used for result validation). A cell in CPT represents CP
+// relations between two fields and consists of a pair of numbers: count and
+// distance. When updating a cell, call updatePairByMerging().
+void CloseProximityBuilder::updateCPT(FieldNumType Src, FieldNumType Dest,
+                                      ExecutionCountType C, DataBytesType D,
+                                      CloseProximityTableType &CPT) {
+  assert(Src <= NumElements && Dest <= NumElements);
+  if (Src == 0 || Dest == 0 || C < 1e-3 || Src == Dest)
+    // Give up update if from or to a dummy node, or if the count is zero
+    return;
+  assert(!std::isnan(C));
+  if (Src != 0 && Dest != 0) {
+    if (Src > Dest) {
+      std::swap(Src, Dest);
+    }
+
+    DEBUG_WITH_TYPE(
+        DEBUG_TYPE_CPG,
+        dbgs() << " Before update: "
+               << "(" << DEBUG_PRINT_COUNT(CPT[Src - 1][Dest - 1].first) << ","
+               << DEBUG_PRINT_DIST(CPT[Src - 1][Dest - 1].second) << ")\n");
+    assert(CPT[Src - 1][Dest - 1].first + C != 0);
+    updatePairByMerging(CPT[Src - 1][Dest - 1].first,
+                        CPT[Src - 1][Dest - 1].second, C, D);
+    if (std::isnan(CPT[Src - 1][Dest - 1].first) ||
+        std::isnan(CPT[Src - 1][Dest - 1].second)) {
+      errs() << "Update CPG between " << Src << " and " << Dest
+             << " with count " << DEBUG_PRINT_COUNT(C) << " and distance "
+             << DEBUG_PRINT_DIST(D) << "\n";
+    }
+    DEBUG_WITH_TYPE(
+        DEBUG_TYPE_CPG,
+        dbgs() << " After update: "
+               << "(" << DEBUG_PRINT_COUNT(CPT[Src - 1][Dest - 1].first) << ","
+               << DEBUG_PRINT_DIST(CPT[Src - 1][Dest - 1].second) << ")\n");
+  }
+}
+
+// Wrapper function to update a cell in CloseProximityTable
+void CloseProximityBuilder::updateCPG(FieldNumType Src, FieldNumType Dest,
+                                      ExecutionCountType C, DataBytesType D) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_CPG,
+                  dbgs() << "Will update CPG between " << Src << " and " << Dest
+                         << " with count " << DEBUG_PRINT_COUNT(C)
+                         << " and distance " << DEBUG_PRINT_DIST(D) << ":\n");
+  updateCPT(Src, Dest, C, D, CloseProximityTable);
```

Comment:
The two functions are used to update CPG and GoldCPG resepectively. It's also a bad legacy from previous version because I want to print different messages for different CPG in debug to check correctness so I used two different functions instead of one... If I removed the function, then I need to check which CPG I am updating before printing a debug message...